x-pack/plugins/apm/server/lib/services/get_services/get_service_transaction_stats.ts

```diff
@@ -0,0 +1,201 @@
+ */
+
+import {
+  AGENT_NAME,
+  SERVICE_ENVIRONMENT,
+  SERVICE_NAME,
+  TRANSACTION_TYPE,
+} from '../../../../common/elasticsearch_fieldnames';
+import {
+  TRANSACTION_PAGE_LOAD,
+  TRANSACTION_REQUEST,
+} from '../../../../common/transaction_types';
+import { rangeFilter } from '../../../../common/utils/range_filter';
+import { AgentName } from '../../../../typings/es_schemas/ui/fields/agent';
+import {
+  getDocumentTypeFilterForAggregatedTransactions,
+  getProcessorEventForAggregatedTransactions,
+  getTransactionDurationFieldForAggregatedTransactions,
+} from '../../helpers/aggregated_transactions';
+import { getBucketSize } from '../../helpers/get_bucket_size';
+import {
+  calculateTransactionErrorPercentage,
+  getOutcomeAggregation,
+} from '../../helpers/transaction_error_rate';
+import { ServicesItemsSetup } from './get_services_items';
+
+interface AggregationParams {
+  setup: ServicesItemsSetup;
+  searchAggregatedTransactions: boolean;
+}
+
+const MAX_NUMBER_OF_SERVICES = 500;
+
+function calculateAvgDuration({
+  value,
+  deltaAsMinutes,
+}: {
+  value: number;
+  deltaAsMinutes: number;
+}) {
+  return value / deltaAsMinutes;
+}
+
+export async function getServiceTransactionStats({
+  setup,
+  searchAggregatedTransactions,
+}: AggregationParams) {
+  const { apmEventClient, start, end, esFilter } = setup;
+
+  const count = {
+    value_count: {
+      field: getTransactionDurationFieldForAggregatedTransactions(
+        searchAggregatedTransactions
+      ),
+    },
+  };
+
+  const outcomes = getOutcomeAggregation({ searchAggregatedTransactions });
+
+  const metrics = {
+    count,
+    avg_duration: {
+      avg: {
+        field: getTransactionDurationFieldForAggregatedTransactions(
+          searchAggregatedTransactions
+        ),
+      },
+    },
+    outcomes,
+  };
+
+  const response = await apmEventClient.search({
+    apm: {
+      events: [
+        getProcessorEventForAggregatedTransactions(
+          searchAggregatedTransactions
+        ),
+      ],
+    },
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { range: rangeFilter(start, end) },
+            ...esFilter,
+            ...getDocumentTypeFilterForAggregatedTransactions(
+              searchAggregatedTransactions
+            ),
+          ],
+        },
+      },
+      aggs: {
+        services: {
+          terms: {
+            field: SERVICE_NAME,
+            size: MAX_NUMBER_OF_SERVICES,
+          },
+          aggs: {
+            transactionType: {
+              terms: {
+                field: TRANSACTION_TYPE,
+                order: { count: 'desc' },
+              },
+              aggs: {
+                ...metrics,
+                environments: {
+                  terms: {
+                    field: SERVICE_ENVIRONMENT,
+                    missing: '',
+                  },
+                },
+                agentName: {
+                  top_hits: {
```

Comment:
can you change this to `top_metrics`? (and measure the performance impact). edge was just updated to an ES snapshot that supports keywords for `top_metrics`. It will be available in the 7.11 ES release. I expect it to have a significant impact.

Reply:
@dgieselaar if that's ok for you I'd rather doing more changes to this query on a following PR.